### PR TITLE
Sema: validate class layout when validating an extension

### PR DESCRIPTION
**Pull request by shajrawi**
rdar://problem/36704036

This reverts a change done in 9b7f3f73d2c8368c5db0ba0f978ad960c39e9970, in a rare case, shown in rdar://problem/36704036, When doing emitVirtualMethodValue in IRGen we can miss the correct offset because we have a bad ClassDecl, The ClassDecl is junk / null because of the before mentioned commit.

The radar contains a sample ClassDecl and IR that shows the problem, unfortunately, I am unable to create a small test-case the reproduces this issue: the layout is wrong only in a single method call in a large project. Reducing the project to something of a reasonable size / removing most of the code “hides” the problem.

Basically, if we look at the ClassDecl, we see that the type for some vars is null type (v.s. the correct type), the interface type is missing and we are missing a few accessor_decl  and parameter_list

### CODE REVIEW

**Comment by shajrawi** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
Can you file a radar and assign it to me to investigate the issue and fix it properly? This change was a significant compile time speedup in non-WMO mode, and it should be easy to reduce a test case for it -- it just requires multiple files.

**Comment by DougGregor** [comment]
This is going to have a nontrivial impact on compile-time performance, and is a very heavy hammer for such a bug. It would be better if we could determine why the method in question failed to request the class's layout when it should have.

**Comment by shajrawi** [comment]
I uploaded new AST to the radar - will close this PR
